include/rmm/mr/device/limiting_resource_adaptor.hpp

```diff
@@ -127,17 +129,19 @@ class limiting_resource_adaptor final : public device_memory_resource {
    */
   void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
   {
-    void* p = nullptr;
-
-    std::size_t proposed_size = rmm::detail::align_up(bytes, allocation_alignment_);
-    if (proposed_size + allocated_bytes_ <= allocation_limit_) {
-      p = upstream_->allocate(bytes, stream);
-      allocated_bytes_ += proposed_size;
-    } else {
-      throw rmm::bad_alloc{"Exceeded memory limit"};
+    auto const proposed_size = rmm::detail::align_up(bytes, allocation_alignment_);
+    auto const old           = allocated_bytes_.fetch_add(proposed_size);
+    if (old + proposed_size <= allocation_limit_) {
+      try {
+        return upstream_->allocate(bytes, stream);
+      } catch (...) {
+        allocated_bytes_ -= proposed_size;
```

Comment:
Why doesn't this need to be .fetch_sub?

Reply:
`operator-=(x)` is equivalent to `fetch_sub(x)`. `fetch_add` was used above to retrieve the old value. The old value isn't needed here, thus the less verbose `operator-=` could be used.